[wasm] Wasm.Build.Tests - fixes for tests failing on CI

radical opened the pull request:
- Fix an issue with stdout/stderr streams not being flushed correctly on nodejs.
- Fix an issue with completely reading stdout/stderr for `xharness`
- And remove dependence on xharness' std* streams, and instead use the emitted `wasm-console.log`, for non-browser cases
  - note: this is still correctly testing the *app* output, as xharness reads that, and writes to `wasm-console.log`
- And copy sdk for workload testing only when running on CI.

Fixes https://github.com/dotnet/runtime/issues/70675

pavelsavara commented:
We will need to update this PR after https://github.com/dotnet/runtime/pull/70746

radical commented:
/azp run runtime-wasm

radical commented:
/azp run runtime-wasm

radical commented:
/azp run runtime-wasm

radical committed:
[wasm][nodejs] Ensure that stdout/stderr have been flushed out before exiting

When the results xml is large, and we are writing the base64
representation in one line, `node` can exit before all the output gets
flushed out. This results in xharness getting an incomplete
`STARTRESULTXML <len> <base64> ... ` with missing `ENDRESULTXML`, thus
no `testResults.xml` is generated.

This can be seen in the case of `Microsoft.Extensions.Primitives.Tests`
which has xml ~140KB, and `System.Memory.Tests` which has a xml ~13MB.

So, wait for the two streams to be flushed out, with a timeout of 3secs.

radical committed:
[wasm] Wasm.Build.Tests: improve reading output from processes

- Fix to call `WaitForExit()` once `WaitForExit(int)` returns, which
  ensures that all the async handlers have been run.

- Also, for non-browser xharness runs use the emitted `wasm-console.log`
  as the output, so we don't depend on xharness' stdout.

radical committed:
Wasm.Build.Tests: Run the crypto test only with browser. Other cases are covered in library tests

radical committed:
Enable WasmTemplateTests.ConsolePublishAndRun

Fixes https://github.com/dotnet/runtime/issues/70675

radical commented:
/azp run runtime-wasm

radical committed:
Wasm.Build.Tests: avoid unncessary copy when building locally

Copy sdk for testing workloads only on CI.

radical commented:
/azp run runtime-wasm

radical commented:
this has a timeout right now for flushing the streams, but maybe that should be removed?

kg commented on src/mono/wasm/templates/templates/console/app-support.mjs:
```diff
@@ -47,11 +47,26 @@ function set_exit_code(exit_code, reason) {
-        App.INTERNAL.mono_wasm_exit(exit_code);
+        let _flush = function(_stream) {
+             return new Promise((resolve, reject) => {
+                 setTimeout(reject, 30000);
```
maybe pass a 'timed out' message to the rejection handler here

kg commented on src/mono/wasm/templates/templates/console/app-support.mjs:
```diff
@@ -47,11 +47,26 @@ function set_exit_code(exit_code, reason) {
+             return new Promise((resolve, reject) => {
+                 setTimeout(reject, 30000);
+                 _stream.write('', function() { resolve () });
+                 _stream.on('error', (error) => reject(error));
```
shouldn't you do this before the write instead of after, since .write could cause an error?

radical replied:
Should we remove the timeout completely for console template, IOW, user apps?

the other usage of the timeout is in tests, which should be useful.

kg replied:
I think removing it for apps is probably reasonable

kg commented on src/mono/wasm/test-main.js:
```diff
@@ -105,26 +105,23 @@ function set_exit_code(exit_code, reason) {
-                });
+            let _flush = function(_stream) {
+                 return new Promise((resolve, reject) => {
+                     setTimeout(reject, 30000);
```
same feedback as for app-support.mjs applies here

kg commented on src/tests/BuildWasmApps/Wasm.Build.Tests/BuildTestBase.cs:
```diff
@@ -230,6 +234,12 @@ public BuildTestBase(ITestOutputHelper output, SharedBuildPerTestClassFixture bu
+            {
+                string wasmConsolePath = Path.Combine(testLogPath, "wasm-console.log");
+                if (File.Exists(wasmConsolePath))
+                    output = File.ReadAllText(wasmConsolePath);
```
if it doesn't exist, maybe emit a warning to specify that we looked for the log and it wasn't there?

kg replied:
this read could also hit a sharing violation if the test process has it locked too, i think? i'm not sure. might be worth having a try-catch around it

kg reviewed (changes requested):
This looks fine to me other than the bits of feedback I provided, changes not mandatory

radical committed:
Address feedback from @kg

radical commented:
/azp run runtime-wasm

radical committed:
Remove timeout on flushing stdout/stderr streams, for console templates, IOW, user apps

radical commented:
/azp run runtime-wasm

radical commented:
Chrome debugger test failure is intermittent, and unrelated.
